# Problem 2
We need to predict probable uptake or purchase count of the products given in "Q2-new-products.csv"

## Import necessary libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
import boto3
import pandas as pd
from io import BytesIO, StringIO

# Initialize a boto3 client with your AWS credentials
# Access key ID & Secret access key shared in your mail
s3_client = boto3.client(
    's3',
    aws_access_key_id='AKIAZI2LHFBCPY22756I',
    aws_secret_access_key='3ndd8vA95ikn39dd0dlNiIUJ6apF9mLu7RlsIWz5'
)

## Reading Data

In [3]:
### Read  CSV file from S3
# Specify your  AWS S3 bucket  name
bucket_name = 'team-err0r404-datathon-3.0'
DATA_CONSUMTION = 'TBL_DATA_CONSUMPTION.csv'
TBL_CUSTOMER_PROFILE = 'TBL_CUSTOMER_PROFILE.csv'
TBL_PACK_PURCHASE = 'TBL_PACK_PURCHASE.csv'
TBL_PRODUCT_CATALOGUE = 'TBL_PRODUCT_CATALOGUE.csv'

Q2_NEW_PRODUCTS = 'Q2-new-products.csv'


# Get the object from the bucket
response1 = s3_client.get_object(Bucket=bucket_name, Key=DATA_CONSUMTION)
response2 = s3_client.get_object(Bucket=bucket_name, Key=TBL_CUSTOMER_PROFILE)
response3 = s3_client.get_object(Bucket=bucket_name, Key=TBL_PACK_PURCHASE)
response4 = s3_client.get_object(Bucket=bucket_name, Key=TBL_PRODUCT_CATALOGUE)

response6 = s3_client.get_object(Bucket=bucket_name, Key=Q2_NEW_PRODUCTS)

# Read the object (which is in bytes) into a DataFrame
data_con_df = pd.read_csv(BytesIO(response1['Body'].read()))
cus_prof_df =  pd.read_csv(BytesIO(response2['Body'].read()))
pack_purc_df =  pd.read_csv(BytesIO(response3['Body'].read()))
prod_cat_df =  pd.read_csv(BytesIO(response4['Body'].read()))

q2_new_products_df = pd.read_csv(BytesIO(response6['Body'].read()))

## Utility Function

In [4]:
def analyze(df):
    print(f"Number of rows: {len(df)}")    
    print(f"Number of columns: {len(df.columns)}")
    
    info_df = pd.DataFrame({
        "NUNIQUE": df.nunique(), # shows the number of unique entries in each column
        "DTYPE": df.dtypes, # shows datatypes of each column's entries
        "NaN Values": df.isna().sum() # shows if there's any missing value in each column
    })
    
    print()
    print(info_df)

To begin with, we start exploring all given datasets and their properties.

We also perform data preprocessing when necessary.

### Deal with Customer Profile Table

In [6]:
cus_prof_df.head(10)

CUSTOMER_ID SERVICE_TYPE  FIRST_CONNECTION_DATE  CONNECTION_TYPE  \
0  CD1F7653F6BE80     Standard               20191231      Fiber Optic   
1  CD1F768EE45143     Standard               20211107  Cable Broadband   
2  CD1F769AD050BF     Standard               20230112      Fiber Optic   
3  CD1F769C960283     Standard               20220611  Cable Broadband   
4  CD1F77E09E8A7A     Standard               20220920      Fiber Optic   
5  CD1F77EA9ABC61     Standard               20230322   Fixed Wireless   
6  CD1F7828E28D37     Standard               20201015      Fiber Optic   
7  CD1F786814B8AF     Standard               20220826  Cable Broadband   
8  CD1F78EE194EBB     Standard               20230228   Fixed Wireless   
9  CD1F796F6478D6     Standard               20230729   Fixed Wireless   

  ROUTER_CATEGORY  STREAMING_VOL_PERCENTAGE  WEB_BROWSING_VOL_PERCENTAGE  \
0         Low End                   70.0205                      14.6609   
1        High End                   72.1494                      11.6056   
2        High End                   47.3220                      18.2921   
3         Low End                   41.2983                      26.2252   
4         Low End                   32.0709                      59.0923   
5        High End                   56.1554                      14.4754   
6         Low End                   41.9904                      23.4721   
7         Low End                   75.8976                       6.9042   
8         Low End                   76.7676                       7.0320   
9        High End                   36.6942                       8.0083   

   IM_VOICE_CALL_VOL_PERCENTAGE  SNS_VOL_PERCENTAGE  \
0                        3.6914              8.5530   
1                        3.2767              1.6749   
2                        3.2387             20.0176   
3                        4.2466             22.3891   
4                        3.6369              4.9594   
5                       20.4694              4.2157   
6                        1.6491             26.9816   
7                        7.4089              8.5071   
8                        6.8712              7.8469   
9                       43.6567              8.0685   

   FILE_ACCESS_VOL_PERCENTAGE  GAME_VOL_PERCENTAGE  NAVIGATION_VOL_PERCENTAGE  \
0                      2.1653               0.0297                     0.0376   
1                     10.2768               0.0000                     0.2096   
2                      3.9060               0.0180                     0.0540   
3                      0.8242               0.0009                     0.0061   
4                      0.0000               0.0000                     0.0000   
5                      3.5471               0.0000                     0.0088   
6                      5.3547               0.0378                     0.0196   
7                      0.8801               0.0000                     0.0078   
8                      0.8172               0.0002                     0.0067   
9                      1.6959               0.0000                     0.0168   

   EMAIL_VOL_PERCENTAGE  OTHERS_VOL_PERCENTAGE  
0                0.0000                 0.8416  
1                0.0000                 0.8069  
2                0.0506                 7.1010  
3                0.0000                 5.0095  
4                0.0000                 0.2405  
5                0.0000                 1.1282  
6                0.0000                 0.4947  
7                0.0000                 0.3944  
8                0.0000                 0.6583  
9                0.0000                 1.8595

In [7]:
analyze(cus_prof_df)

Number of rows: 342643
Number of columns: 14

                              NUNIQUE    DTYPE  NaN Values
CUSTOMER_ID                    342643   object           0
SERVICE_TYPE                        2   object           0
FIRST_CONNECTION_DATE            7637    int64           0
CONNECTION_TYPE                     3   object           0
ROUTER_CATEGORY                     2   object           0
STREAMING_VOL_PERCENTAGE       263200  float64        1946
WEB_BROWSING_VOL_PERCENTAGE    225945  float64        1946
IM_VOICE_CALL_VOL_PERCENTAGE   166005  float64        1946
SNS_VOL_PERCENTAGE             153181  float64        1946
FILE_ACCESS_VOL_PERCENTAGE     108813  float64        1946
GAME_VOL_PERCENTAGE             20037  float64        1946
NAVIGATION_VOL_PERCENTAGE       11475  float64        1946
EMAIL_VOL_PERCENTAGE             3722  float64        1946
OTHERS_VOL_PERCENTAGE           81795  float64        1946


In [8]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

As we can see from above analysis,

* The date formats of `FIRST_CONNECTION_DATE` are in `20191231` format that needs to be converted to machine learnable `2019-12-31` format.
* All the `PERCENTAGE`-columns have same amount of missing values that needs to be dealt with.
* Since `SERVICE_TYPE`, `CONNECTION_TYPE` and `ROUTER_CATEGORY` are all categorical columns, they can be encoded for efficiency.

In [9]:
# Convert the format of dates to datetime
cus_prof_df['FIRST_CONNECTION_DATE'] = pd.to_datetime(cus_prof_df['FIRST_CONNECTION_DATE'], format='%Y%m%d')

In [10]:
# Encode the categorical columns

cus_prof_df['SERVICE_TYPE'] = le.fit_transform(cus_prof_df['SERVICE_TYPE'])
cus_prof_df['CONNECTION_TYPE'] = le.fit_transform(cus_prof_df['CONNECTION_TYPE'])
cus_prof_df['ROUTER_CATEGORY'] = le.fit_transform(cus_prof_df['ROUTER_CATEGORY'])

In [11]:
cus_prof_df.head()

CUSTOMER_ID  SERVICE_TYPE FIRST_CONNECTION_DATE  CONNECTION_TYPE  \
0  CD1F7653F6BE80             1            2019-12-31                1   
1  CD1F768EE45143             1            2021-11-07                0   
2  CD1F769AD050BF             1            2023-01-12                1   
3  CD1F769C960283             1            2022-06-11                0   
4  CD1F77E09E8A7A             1            2022-09-20                1   

   ROUTER_CATEGORY  STREAMING_VOL_PERCENTAGE  WEB_BROWSING_VOL_PERCENTAGE  \
0                1                   70.0205                      14.6609   
1                0                   72.1494                      11.6056   
2                0                   47.3220                      18.2921   
3                1                   41.2983                      26.2252   
4                1                   32.0709                      59.0923   

   IM_VOICE_CALL_VOL_PERCENTAGE  SNS_VOL_PERCENTAGE  \
0                        3.6914              8.5530   
1                        3.2767              1.6749   
2                        3.2387             20.0176   
3                        4.2466             22.3891   
4                        3.6369              4.9594   

   FILE_ACCESS_VOL_PERCENTAGE  GAME_VOL_PERCENTAGE  NAVIGATION_VOL_PERCENTAGE  \
0                      2.1653               0.0297                     0.0376   
1                     10.2768               0.0000                     0.2096   
2                      3.9060               0.0180                     0.0540   
3                      0.8242               0.0009                     0.0061   
4                      0.0000               0.0000                     0.0000   

   EMAIL_VOL_PERCENTAGE  OTHERS_VOL_PERCENTAGE  
0                0.0000                 0.8416  
1                0.0000                 0.8069  
2                0.0506                 7.1010  
3                0.0000                 5.0095  
4                0.0000                 0.2405

## Deal with Product Catalogue Table

In [12]:
prod_cat_df.head(10)

PACK_ID SERVICE_TYPE   PAYMENT_METHOD  PACK_PRICE  VALIDITY  DATA_VOL_GB  \
0        1     Standard  Digital Wallets        1298        30        320.0   
1        2     Standard             Cash         328        30         24.0   
2        3      Premium     Mobile Walet         176         3         24.0   
3        4      Premium             Cash         558        15         64.0   
4        5      Premium             Cash         300         7         20.0   
5        6      Premium     Mobile Walet         196         3         28.0   
6        7      Premium     Mobile Walet         898        30         44.0   
7        8      Premium      Credit Card         990        30        160.0   
8        9      Premium  Digital Wallets         136         3         16.0   
9       10      Premium             Cash         440         7         80.0   

  PRODUCT_STATUS  
0    Discontinue  
1       Continue  
2    Discontinue  
3    Discontinue  
4            New  
5    Discontinue  
6    Discontinue  
7    Discontinue  
8    Discontinue  
9            New

In [13]:
analyze(prod_cat_df)

Number of rows: 2049
Number of columns: 7

                NUNIQUE    DTYPE  NaN Values
PACK_ID            2049    int64           0
SERVICE_TYPE          2   object           0
PAYMENT_METHOD        4   object           0
PACK_PRICE          291    int64           0
VALIDITY              4    int64           0
DATA_VOL_GB         163  float64           0
PRODUCT_STATUS        3   object           0


As we discussed above, 

`SERVICE_TYPE`, `PAYMENT_TYPE` and `PRODUCT_STATUS` are categorical columns, so we encode them.

In [14]:
# Encode the categorical columns

prod_cat_df['SERVICE_TYPE'] = le.fit_transform(prod_cat_df['SERVICE_TYPE'])
prod_cat_df['PAYMENT_METHOD'] = le.fit_transform(prod_cat_df['PAYMENT_METHOD'])
prod_cat_df['PRODUCT_STATUS'] = le.fit_transform(prod_cat_df['PRODUCT_STATUS'])

In [15]:
prod_cat_df.head()

PACK_ID  SERVICE_TYPE  PAYMENT_METHOD  PACK_PRICE  VALIDITY  DATA_VOL_GB  \
0        1             1               2        1298        30        320.0   
1        2             1               0         328        30         24.0   
2        3             0               3         176         3         24.0   
3        4             0               0         558        15         64.0   
4        5             0               0         300         7         20.0   

   PRODUCT_STATUS  
0               1  
1               0  
2               1  
3               1  
4               2

## Deal with Pack Purchase Table

In [16]:
pack_purc_df.head(10)

EVENT_DATE     CUSTOMER_ID  PACK_ID  HITS
0    20190703  CD20EDFC8000A7     1756     1
1    20190522  CD1F932E0A89C1     1524     1
2    20190522  CD201715F62164     1524     1
3    20190623  CD1F7B4DD83C5B     1524     1
4    20190523  CD1F7BD12742B8     1524     1
5    20190522  CD1F89B2CE86FB     1524     1
6    20190522  CD1F8F6C200BA1     1524     1
7    20190624  CD1F926FE0B501     1524     1
8    20190621  CD1F931C4A1659     1524     1
9    20190614  CD1F93A1305BA6     1524     1

In [17]:
analyze(pack_purc_df)

Number of rows: 1208710
Number of columns: 4

             NUNIQUE   DTYPE  NaN Values
EVENT_DATE       131   int64           0
CUSTOMER_ID   342635  object           0
PACK_ID         1522   int64           0
HITS              10   int64           0


Similar to previous situation, we must convert the dates to machine learnable format.

In [18]:
# Convert the format of dates to datetime

pack_purc_df['EVENT_DATE'] = pd.to_datetime(pack_purc_df['EVENT_DATE'], format='%Y%m%d')

In [19]:
pack_purc_df.head()

EVENT_DATE     CUSTOMER_ID  PACK_ID  HITS
0 2019-07-03  CD20EDFC8000A7     1756     1
1 2019-05-22  CD1F932E0A89C1     1524     1
2 2019-05-22  CD201715F62164     1524     1
3 2019-06-23  CD1F7B4DD83C5B     1524     1
4 2019-05-23  CD1F7BD12742B8     1524     1

## Deal with Data Consumption Table

In [20]:
data_con_df.head(10)

EVENT_DATE     CUSTOMER_ID  TOTAL_VOLUME_GB  VOLUME_GB_IN_HIGH_BANDWIDTH
0    20190914  CD1F7653F6BE80          14.2695                      14.2695
1    20190916  CD1F7653F6BE80          22.2891                      22.2891
2    20190716  CD1F7653F6BE80          26.3086                      26.3086
3    20190906  CD1F7653F6BE80           6.8477                       6.7813
4    20190921  CD1F7653F6BE80          15.1406                      15.1406
5    20190816  CD1F7653F6BE80           1.8320                       1.8047
6    20190710  CD1F7653F6BE80          18.8086                      18.8086
7    20190911  CD1F7653F6BE80           6.3594                       6.3594
8    20190817  CD1F7653F6BE80           4.2305                       4.0938
9    20190823  CD1F7653F6BE80           3.6797                       3.5977

In [21]:
analyze(data_con_df)

Number of rows: 23375373
Number of columns: 4

                             NUNIQUE    DTYPE  NaN Values
EVENT_DATE                       183    int64           0
CUSTOMER_ID                   333503   object           0
TOTAL_VOLUME_GB                23484  float64           0
VOLUME_GB_IN_HIGH_BANDWIDTH    23482  float64           0


In [22]:
# Convert the format of dates to datetime

data_con_df['EVENT_DATE'] = pd.to_datetime(data_con_df['EVENT_DATE'], format='%Y%m%d')

In [23]:
data_con_df.head()

EVENT_DATE     CUSTOMER_ID  TOTAL_VOLUME_GB  VOLUME_GB_IN_HIGH_BANDWIDTH
0 2019-09-14  CD1F7653F6BE80          14.2695                      14.2695
1 2019-09-16  CD1F7653F6BE80          22.2891                      22.2891
2 2019-07-16  CD1F7653F6BE80          26.3086                      26.3086
3 2019-09-06  CD1F7653F6BE80           6.8477                       6.7813
4 2019-09-21  CD1F7653F6BE80          15.1406                      15.1406

# Solving Problem 2 

At first, let's check if all `PACK_ID` of Product Catalogue are present in Package Purchased.

In [24]:
prod_cat_df["PACK_ID"].isin(pack_purc_df["PACK_ID"])

0        True
1        True
2        True
3        True
4       False
        ...  
2044     True
2045     True
2046     True
2047    False
2048     True
Name: PACK_ID, Length: 2049, dtype: bool

In [25]:
## Q2 -> New Products

q2_new_products_df

PACK_ID
0       352
1      1735
2       179
3      1434
4       426
5        77
6      1856
7       150
8      1563
9       801
10     1424
11     1587
12     1576
13       90
14      404
15       29
16     1768
17     1705
18       74
19      544

Since, we need to estimate the "Probable Uptake" or "Purchase Count" of the new given products, we look for datasets that has `PACK_ID` and `HITS`.

We can see that, `prod_cat_df` has `PACK_ID` that refers to the information about the packages. On the other hand, `pack_purc_df` has `PACK_ID` and `HITS` that refer to the packages purchased by the customers in the given `EVENT_DATE`.

That's why we ought to merge them.

In [26]:
merged_df = pd.merge(prod_cat_df, pack_purc_df, on="PACK_ID", how="left")

In [27]:
merged_df.head()

PACK_ID  SERVICE_TYPE  PAYMENT_METHOD  PACK_PRICE  VALIDITY  DATA_VOL_GB  \
0        1             1               2        1298        30        320.0   
1        1             1               2        1298        30        320.0   
2        1             1               2        1298        30        320.0   
3        1             1               2        1298        30        320.0   
4        1             1               2        1298        30        320.0   

   PRODUCT_STATUS EVENT_DATE     CUSTOMER_ID  HITS  
0               1 2019-08-28  CD1F881845A6B5   1.0  
1               1 2019-08-29  CD1F89A8749D3B   1.0  
2               1 2019-09-06  CD1FCBEBCAA71F   1.0  
3               1 2019-09-10  CD1F7DA0BA986A   1.0  
4               1 2019-08-11  CD1F86EA1AFDB9   1.0

In [28]:
analyze(merged_df)

Number of rows: 1209237
Number of columns: 10

                NUNIQUE           DTYPE  NaN Values
PACK_ID            2049           int64           0
SERVICE_TYPE          2           int64           0
PAYMENT_METHOD        4           int64           0
PACK_PRICE          291           int64           0
VALIDITY              4           int64           0
DATA_VOL_GB         163         float64           0
PRODUCT_STATUS        3           int64           0
EVENT_DATE          131  datetime64[ns]         527
CUSTOMER_ID      342635          object         527
HITS                 10         float64         527


Now, we need to keep in mind that we are trying to estimate the `HITS` of given `PACK_ID` for the next month (October).

Hence, we should separate `EVENT_DATE`'s `YEAR`, `MONTH` and `DAY`.

In [29]:
merged_df['YEAR'] = merged_df['EVENT_DATE'].dt.year
merged_df['MONTH'] = merged_df['EVENT_DATE'].dt.month
merged_df['DAY'] = merged_df['EVENT_DATE'].dt.day

In [30]:
# We no longer need EVENT_DATE, so we drop it.

merged_df.drop(columns=["EVENT_DATE"], inplace=True)
analyze(merged_df)

Number of rows: 1209237
Number of columns: 12

                NUNIQUE    DTYPE  NaN Values
PACK_ID            2049    int64           0
SERVICE_TYPE          2    int64           0
PAYMENT_METHOD        4    int64           0
PACK_PRICE          291    int64           0
VALIDITY              4    int64           0
DATA_VOL_GB         163  float64           0
PRODUCT_STATUS        3    int64           0
CUSTOMER_ID      342635   object         527
HITS                 10  float64         527
YEAR                  1  float64         527
MONTH                 5  float64         527
DAY                  31  float64         527


We also don't require `CUSTOMER_ID`, `DAY` and `YEAR` since Customer IDs don't play any role here as well as the days and years because we focus on predicting based on the `MONTH`.

In [31]:
# Drop non-required columns

merged_df.drop(columns=["CUSTOMER_ID", "DAY", "YEAR"], inplace=True)
merged_df.head()

PACK_ID  SERVICE_TYPE  PAYMENT_METHOD  PACK_PRICE  VALIDITY  DATA_VOL_GB  \
0        1             1               2        1298        30        320.0   
1        1             1               2        1298        30        320.0   
2        1             1               2        1298        30        320.0   
3        1             1               2        1298        30        320.0   
4        1             1               2        1298        30        320.0   

   PRODUCT_STATUS  HITS  MONTH  
0               1   1.0    8.0  
1               1   1.0    8.0  
2               1   1.0    9.0  
3               1   1.0    9.0  
4               1   1.0    8.0

In [32]:
analyze(merged_df)

Number of rows: 1209237
Number of columns: 9

                NUNIQUE    DTYPE  NaN Values
PACK_ID            2049    int64           0
SERVICE_TYPE          2    int64           0
PAYMENT_METHOD        4    int64           0
PACK_PRICE          291    int64           0
VALIDITY              4    int64           0
DATA_VOL_GB         163  float64           0
PRODUCT_STATUS        3    int64           0
HITS                 10  float64         527
MONTH                 5  float64         527


As we can see from above analysis, some `HITS` and `MONTH` have missing values. We fill them both with Zero (0) which demotes "no hit" for `HITS` and dummy value for `MONTH`.

In [33]:
# We fill accordingly

merged_df.fillna({'HITS': 0, 'MONTH': 0}, inplace=True)
merged_df = merged_df[["PACK_ID", "MONTH", "HITS"]]

In [34]:
analyze(merged_df)

Number of rows: 1209237
Number of columns: 3

         NUNIQUE    DTYPE  NaN Values
PACK_ID     2049    int64           0
MONTH          6  float64           0
HITS          11  float64           0


## Model Training

In [35]:
# Splitting the final_merged_df into features (X) and target (y)
X = merged_df.drop(columns=['HITS'])
y = merged_df['HITS']

In [36]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor

In [37]:
# Splitting the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize the model
model = RandomForestRegressor(n_estimators=100, random_state=42)

# Training the model
model.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

In [38]:
import math
from sklearn.metrics import mean_squared_error

# Making predictions
predictions = model.predict(X_test)

# Evaluating the model
mse = mean_squared_error(y_test, predictions)
print(f"RMSE: {math.sqrt(mse)}")

RMSE: 0.07553056268413529


In [39]:
analyze(X_test)

Number of rows: 241848
Number of columns: 2

         NUNIQUE    DTYPE  NaN Values
PACK_ID     1385    int64           0
MONTH          6  float64           0


In [40]:
q2_new_products_df.head()

PACK_ID
0      352
1     1735
2      179
3     1434
4      426

Now for prediction purpose, we add a column `MONTH` to `q2_new_products_df` filled with `10` denoting the month "October".

In [41]:
q2_new_products_df["MONTH"] = 10

In [42]:
q2_new_products_df["HITS"] = model.predict(q2_new_products_df)

In [43]:
q2_new_products_df.head()

PACK_ID  MONTH      HITS
0      352     10  1.000000
1     1735     10  1.020000
2      179     10  1.003163
3     1434     10  1.000000
4      426     10  1.590032

Since we have successfully predicted `HITS` for given `PACK_ID`, we drop `MONTH` column as we don't need that anymore.

In [45]:
q2_new_products_df.drop(columns=["MONTH"], inplace=True)

## Write the submission file

In [47]:
### write CSV file to S3
write_file_name = 'Q2-submission.csv'

# Convert the dataframe to CSV
csv_buffer = StringIO()
q2_new_products_df.to_csv(csv_buffer, index = False)

# Write the CSV data to S3
s3_client.put_object(Bucket=bucket_name, Key=write_file_name, Body=csv_buffer.getvalue())

{'ResponseMetadata': {'RequestId': '9G3CSBD32W8ZJMZP',
  'HostId': 'jTX1O/n4XVUBltFz2ku2PaO7mMyLIQC+nfJUfBbLlX4W/xDNcnKpXVhXfuOunGIg2e40VEnJy30=',
  'HTTPStatusCode': 200,
  'HTTPHeaders': {'x-amz-id-2': 'jTX1O/n4XVUBltFz2ku2PaO7mMyLIQC+nfJUfBbLlX4W/xDNcnKpXVhXfuOunGIg2e40VEnJy30=',
   'x-amz-request-id': '9G3CSBD32W8ZJMZP',
   'date': 'Sat, 25 May 2024 04:32:28 GMT',
   'x-amz-server-side-encryption': 'AES256',
   'etag': '"3ad2362eef9bc27d601864f29175a7de"',
   'server': 'AmazonS3',
   'content-length': '0'},
  'RetryAttempts': 0},
 'ETag': '"3ad2362eef9bc27d601864f29175a7de"',
 'ServerSideEncryption': 'AES256'}